tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,506 @@
+
+#### JSON Golden Test
+
+Until now, we have managed components that return only strings, but what if our return type is different like a `ByteString` or `Text`? For example, let's analyze the `encodeCountries` function in the `Json.hs` module.
+
+```haskell
+encodeCountries :: [Country] -> ByteString
+encodeCountries = encode
+```
+
+This function returns a ByteString. That means that if we try to assert the output with our `defaultGolden` function,
+
+```haskell
+  describe "encodeCountries" $ do
+   it "encodes a group of Countries into a JSON String " $
+    defaultGolden "json" (encodeCountries countries)
+```
+
+tests won't even compile.
+
+```shell
+$stack test
+• Couldn't match type ‘B.ByteString’ with ‘[Char]’
+      Expected type: String
+        Actual type: B.ByteString
+```
+
+Conveniently, `hspec-golden` exports a `Golden` data type. With this tool we can configure our Golden test by specifying the functions `hspec-golden` should use for our test to work.
+
+```haskell
+data Golden str =
+  Golden {
+    output       :: str,                      --  Output
+    encodePretty :: str -> String,            --  Makes the comparison pretty when the test fails
+    writeToFile  :: FilePath -> str -> IO (), --  How to write into the golden file the file
+    readFromFile :: FilePath -> IO str,       --  How to read the file,
+    testName     :: String,                   --  Test name 
+    directory    :: FilePath                  --  Directory where you write your tests
+  }
+```
+For example, we can define a different directory for storing our golden files or we can define how our test will be read and how the output should be written. Also, the `encodePretty` characteristic determines how the prompt should print a readable output when tests fail. With this data-type we are now able to create a new function to assert our output. Let's start by importing some modules.
+
+```haskell
+module Json.JsonGoldenSpec where
+
+import           Test.Hspec
+import           Test.Hspec.Golden
+import           JSON.Json
+import qualified Data.ByteString.Lazy as B    
+```
+
+Next, based on the `Golden` data type documentation , let's create our assert function.
+
+```haskell
+goldenBytestring :: String -> B.ByteString -> Golden B.ByteString
+goldenBytestring name actualOutput =
+    Golden {
+        output = actualOutput,
+        encodePretty = show,
+        writeToFile = B.writeFile,
+        readFromFile = B.readFile,
+        testName = name,
+        directory = ".otherGolden"
+          
+    }
+```
+
+Let's analyze a little bit our `goldenBytestring` function: 
+
+ - Its type is `Golden ByteString`.
+ - To print our failed tests it makes use of haskell's `show` function. 
+ - To write the bytestring into a file as well as reading it from a file we have used the `writeFile` and `readFile` functions from the Data.ByteString.Lazy module.
+ - We will be saving the output into a different directory called `.otherGolden/`. 
+
```

Comment:
I think it would be clearer if you order the sentences the other way around. For instance "It uses the `show` function to print out failed tests"